src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileStorageService.cs

```diff
@@ -49,58 +50,110 @@ public async Task<IEnumerable<FileInfo>> ListFilesAsync(Guid releaseId, ReleaseF
             var blobContainer = await GetCloudBlobContainer();
+            // Get the paths of the files to delete
+            return await DataPathsForDeletion(blobContainer, releaseId, fileName)
+                .OnSuccess(async (path) =>
+                {
+                    // Delete the data file
+                    return await DeleteFileAsync(blobContainer, path.dataFilePath)
+                        // and the metadata file
+                        .OnSuccess(async () => await DeleteFileAsync(blobContainer, path.metadataFilePath))
+                        // and return the remaining files
+                        .OnSuccess(async () => await ListFilesAsync(releaseId, ReleaseFileTypes.Data));
+                });
+        }
+        
+        private async static Task<Either<ValidationResult, (string dataFilePath, string metadataFilePath)>> DataPathsForDeletion(CloudBlobContainer blobContainer, Guid releaseId, string fileName)
+        {
+            var dataFilePath = AdminReleasePath(releaseId, ReleaseFileTypes.Data, fileName);
             var dataBlob = blobContainer.GetBlockBlobReference(dataFilePath);
             if (!dataBlob.Exists())
             {
                 return ValidationResult(FileNotFound);
             }
-            dataBlob.FetchAttributes(); 
+
+            dataBlob.FetchAttributes();
             if (!dataBlob.Metadata.ContainsKey(MetaFileKey))
             {
                 return ValidationResult(UnableToFindMetadataFileToDelete);
             }
+
             var metaFileName = dataBlob.Metadata[MetaFileKey];
-            // Delete the data 
-            return await DeleteFileAsync(blobContainer, AdminReleasePath(releaseId, ReleaseFileTypes.Data, fileName))
-                // and the metadata
-                .OnSuccess(async () => await DeleteFileAsync(blobContainer, AdminReleasePath(releaseId, ReleaseFileTypes.Data, metaFileName)))
-                // and return the remaining files
-                .OnSuccess(async () => await ListFilesAsync(releaseId, ReleaseFileTypes.Data));
+            var metadataFilePath = AdminReleasePath(releaseId, ReleaseFileTypes.Data, metaFileName);
+            var metaBlob = blobContainer.GetBlockBlobReference(metadataFilePath);
+            if (!metaBlob.Exists())
+            {
+                return ValidationResult(UnableToFindMetadataFileToDelete);
+            }
+
+            return (dataFilePath : dataFilePath, metadataFilePath : metadataFilePath);
         }
 
-        public async Task<Either<ValidationResult, IEnumerable<FileInfo>>> UploadFilesAsync(Guid releaseId, IFormFile file, string name, ReleaseFileTypes type)
+        public async Task<Either<ValidationResult, IEnumerable<FileInfo>>> UploadFilesAsync(Guid releaseId,
+            IFormFile file, string name, ReleaseFileTypes type)
         {
+            if (type == ReleaseFileTypes.Data)
+            {
+                return ValidationResult(CannotUseGenericFunctionToDeleteDataFile);
```

Comment:
Why do we mention deleting here in an upload method?

Reply:
monkey trumpets - I'll change this